internal/execute/tableutil/stream.go

```diff
@@ -0,0 +1,184 @@
+// If the function returns without returning at least one table,
+// then an error will be returned. If the first table that is returned
+// is empty, then this will return an empty table and further buffers
+// will not be used.
+func StreamContext(ctx context.Context, f func(ctx context.Context, fn SendFunc) error) (flux.Table, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	ch := make(chan streamBuffer)
+
+	// Create the send method.
+	send := func(cr flux.ColReader) {
+		select {
+		case ch <- streamBuffer{cr: cr}:
+		case <-ctx.Done():
+			// We could not send the column reader because this was cancelled.
+			cr.Release()
+		}
+	}
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		defer close(ch)
+		if err := f(ctx, send); err != nil {
+			ch <- streamBuffer{err: err}
+		}
+	}()
+
+	select {
+	case sp := <-ch:
+		cr, err := sp.cr, sp.err
+		if cr == nil {
+			if err == nil {
+				err = errors.New(codes.Internal, "empty table stream")
+			}
+			cancel()
+			return nil, err
+		}
+
+		// Retrieve the group key and columns from the column reader.
+		key, cols := cr.Key(), cr.Cols()
+
+		// If the table is empty, signal to the context
+		// that we are not expecting more tables just
+		// in case the implementor does something wrong.
+		// We also release the column reader since we don't need
+		// it anymore and set it to nil.
+		empty := cr.Len() == 0
+		if empty {
+			cancel()
+			cr.Release()
+			cr = nil
+		}
+		return &streamTable{
+			first:  cr,
+			key:    key,
+			cols:   cols,
+			cancel: cancel,
+			ch:     ch,
+			done:   done,
+			empty:  empty,
+		}, nil
+	case <-ctx.Done():
+		cancel()
+		return nil, ctx.Err()
+	}
+}
+
+// streamBuffer is a column reader or error sent
+// from the streaming function.
+type streamBuffer struct {
+	cr  flux.ColReader
+	err error
+}
+
+// streamTable is an implementation of flux.Table
+// that will stream buffers from a column reader.
+type streamTable struct {
+	used   int32
+	first  flux.ColReader
+	key    flux.GroupKey
+	cols   []flux.ColMeta
+	cancel func()
+	ch     <-chan streamBuffer
+	done   <-chan struct{}
+	empty  bool
+}
+
+func (s *streamTable) Key() flux.GroupKey {
+	return s.key
+}
+
+func (s *streamTable) Cols() []flux.ColMeta {
+	return s.cols
+}
+
+func (s *streamTable) Do(f func(flux.ColReader) error) error {
+	if !atomic.CompareAndSwapInt32(&s.used, 0, 1) {
+		return errors.New(codes.Internal, "table already read")
+	}
+
+	// Ensure that we always call cancel to free any resources from
+	// the context after we have completely read the channel.
+	defer s.cancel()
+
+	// If the table is empty, return immediately.
+	// We already released the column reader.
+	if s.empty {
+		return nil
+	}
+
+	// Act on the first column reader that was read.
+	if err := f(s.first); err != nil {
+		s.first.Release()
+		s.first = nil
+		return nil
+	}
+	s.first.Release()
+	s.first = nil
+
+	for sp := range s.ch {
+		cr, err := sp.cr, sp.err
+		if err != nil {
+			return err
```

Comment:
Is this early return going to be a problem anyway for the done channel?

Reply:
I do not think so. The done channel is just used to prevent this from leaving until the streaming function ends. When an error happens, the streaming function is supposed to be done. If the streaming function continues sending things, then the context gets canceled by this return and we do not want to wait for the done channel. We just need to signal to the sender that we're ignoring it now.